job.c

```diff
@@ -34,24 +34,32 @@ store_job(job j)
     all_jobs_used++;
 
     /* accept a load factor of 4 */
-    if (all_jobs_used > (all_jobs_cap << 2)) rehash();
+    if (all_jobs_used > (all_jobs_cap << 2)) rehash(1);
 }
 
 static void
-rehash()
+rehash(int is_upscaling)
 {
     job *old = all_jobs;
     size_t old_cap = all_jobs_cap, old_used = all_jobs_used, i;
+    int old_prime = cur_prime;
 
-    if (cur_prime >= NUM_PRIMES) return;
-    if (hash_table_was_oom) return;
+    if (is_upscaling) {
+        if (cur_prime >= NUM_PRIMES) return;
+        if (hash_table_max_prime <= cur_prime) return;
+        ++cur_prime;
+    }
+    else {
+        if (cur_prime <= 0) return;
+        --cur_prime;
+    }
 
-    all_jobs_cap = primes[++cur_prime];
+    all_jobs_cap = primes[cur_prime];
     all_jobs = calloc(all_jobs_cap, sizeof(job));
     if (!all_jobs) {
         twarnx("Failed to allocate %zu new hash buckets", all_jobs_cap);
-        hash_table_was_oom = 1;
-        --cur_prime;
+        hash_table_max_prime = cur_prime;
```

Comment:
What I said in https://github.com/kr/beanstalkd/pull/295#discussion_r58302039 still applies.

The name `hash_table_max_prime` is incorrect. There is no such thing as a "max prime" or a maximum size for the hashtable. This code branch is concerned with the *out of memory* condition, which can happen for many reasons, including allocation of entirely unrelated objects. This should be put back the way it was.

Additionally, now that we're also shrinking the table, we should clear the oom flag in the other code path. Below this if statement, there should be a line:

```
if (!all_jobs) {
    twarnx(...);
    hash_table_was_oom = 1;
    ...
}
hash_table_was_oom = 0;
```